### PR TITLE
[Code editor] Pins commit SHA, support updating it, and add breadcrumbs

### PR DESCRIPTION
**Pull request by siggisim**
- Pins workspace to a specific commit SHA (pins head commit if not set)
- Support updating pinned commit SHA to head
- Support resolving conflicts when updating commit SHA with a diff viewer / editor
- Adds breadcrumbs to show pinned commit SHA (and gives a UI entry-point for updating)

---

**Version bump**: Patch

### CODE REVIEW

File: enterprise/app/code/BUILD [hunk 1]
```diff
@@ -14,6 +14,7 @@ ts_library(
         "*.tsx",
     ]),
     deps = [
+        "//app/alert",
```
Comment (tylerwilliams) [inline enterprise/app/code/BUILD:17]: eh?

File: enterprise/app/code/code.tsx [hunk 2]
```diff
@@ -56,14 +60,17 @@ export default class CodeComponent extends React.Component<Props> {
...
     treeShaToExpanded: new Map<string, boolean>(),
     treeShaToChildrenMap: new Map<string, any[]>(),
     treeShaToPathMap: new Map<string, string>(),
     fullPathToModelMap: new Map<string, any>(),
+    fullPathToDiffModelMap: new Map<string, any>(),
     originalFileContents: new Map<string, any>(),
     currentFilePath: "",
     changes: new Map<string, string>(),
+    mergeConflicts: new Map<string, string>(),
     pathToIncludeChanges: new Map<string, boolean>(),
```
Comment (tylerwilliams) [inline enterprise/app/code/code.tsx:74]: maps are the most powerful data structure known to man

File: enterprise/app/code/BUILD (hunk 1 above)
Comment (tylerwilliams) [reply enterprise/app/code/BUILD:17]: eh.

**Review by tylerwilliams — APPROVED** [review]
LGTM

File: enterprise/app/code/code.tsx (hunk 2 above)
Comment (siggisim) [reply enterprise/app/code/code.tsx:74]: loves me some maps lol

**Pull request merged**